### DISCUSSION TITLE
bench restana + turbo-http?

### DISCUSSION THREAD

**Discussion by leeoniya** [uNetworking/uWebSockets.js#10] [Q&A] [answered]
hey @alexhultman,

https://medium.com/@kyberneees/building-ultra-fast-rest-apis-with-node-js-and-restana-1d65b0d524b7

looks to be about 2x faster than Node native. (and can do clustering for another 20%).

**Comment by ghost** [comment]
No need to benchmark, I already know it's not close.

**Comment by ghost** [comment, marked as answer]
It's not even ballpark

turbo-http without SSL (I don't even think they have SSL):
```
alexhultman@alexhultman-N56VZ:~/wrk$ ./wrk http://localhost:8080
Running 10s test @ http://localhost:8080
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   220.30us   27.02us   1.60ms   90.43%
    Req/Sec    22.66k     2.05k   45.04k    90.55%
  453423 requests in 10.10s, 32.00MB read
Requests/sec:  44896.51
Transfer/sec:      3.17MB
```

uws:
```
alexhultman@alexhultman-N56VZ:~/wrk$ ./wrk http://localhost:3000
Running 10s test @ http://localhost:3000
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    53.52us   19.67us 488.00us   73.45%
    Req/Sec    91.13k     7.06k  105.67k    60.89%
  1830880 requests in 10.10s, 89.05MB read
Requests/sec: 181274.01
Transfer/sec:      8.82MB
```

uws with SSL:
```
alexhultman@alexhultman-N56VZ:~/wrk$ ./wrk https://localhost:3000
Running 10s test @ https://localhost:3000
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    91.79us  326.94us  17.42ms   99.79%
    Req/Sec    60.19k     8.21k   71.15k    80.20%
  1209457 requests in 10.10s, 58.82MB read
Requests/sec: 119748.92
Transfer/sec:      5.82MB
```

**Comment by leeoniya** [comment]
i wasn't suggesting it would be close to uWSjs (i know what 2x means). was just curious how it stacked up to native node http in your testing. cool to see it roughly on par with Crystal.

thanks!

**Comment by ghost** [comment]
You mean turbo-http? Uws is not comparable with Crystal. Where do you publish these results?

**Comment by leeoniya** [comment]
https://github.com/uNetworking/uWebSockets/blob/master/misc/bigshot_lineup.png

**Comment by ghost** [comment]
Is turbo-http worth putting in? It's not very well known.

**Comment by leeoniya** [comment]
i think it's interesting since it shows you can do faster than Node "native" without compiled addons. if you want to be fair to "what's possible" in current Node, then maybe it's worth adding.

i think it's more interesting than showing how slow the unfinished (and also unknown) Deno is.

**Comment by ghost** [comment]
I don't follow.

> i think it's interesting since it shows you can do faster than Node "native" without compiled addons

This makes no sense. Turbo-http *is* using a native add-on. You can't go faster than "native" without using add-ons unless you manage to build something on net. Server that's somewhat faster. But none of these projects do

**Comment by leeoniya** [comment]
> Turbo-http is using a native add-on.

oops, guess i missed that dependency :man_facepalming:

this whole issue was pointless. sry!

**Comment by ghost** [comment]
It's not pointless for turbo http but it's pointless for everything else (restana, fastify, etc etc etc) since they all rely on node js

**Comment by leeoniya** [comment]
> but it's pointless for everything else (restana, fastify, etc etc etc) since

well, you do have Express in there to show how much it adds to bare Node http. restana does much better, presumably. including a router that does some form or uri interpolation in all cases rather than serving up static routes would be more apples-to-apples?

**Comment by ghost** [comment]
Express is kind of an odd thing to measure, yes. It doesn't fit, it's unnecessary to plot since it cannot be faster than Node.js BUT - it's extremely popular so I just have to add it just because. (Same goes with Socket.IO in the WebSocket benchmark - it makes no sense to have it included but because of its extreme popularity due to the snowballing effect of GitHub stars / NPM downloads I just have to add it). Everything else is not interesting.
